### PR TITLE
Move ABI check to instantiation

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

This is a follow-up to https://github.com/clockworklabs/SpacetimeDB/pull/197.

This fixes an extra case I was worried about but didn't hit before and so had hoped a simple fix above is enough.

Now I've hit it in a different module: because ABI can have zeroes - e.g. 3.0 is encoded as `\00\00\03\00` in the data segment - linker might place it in the beginning of a data segment, in which case `wasm-opt` will trim those prefix \00 bytes and there won't be any single data segment containing the entire version, and the ABI check will still fail.

It's possible to make the `determine_spacetime_abi` account for this case and merge chunks from different data segments when they overlap with version, but I feel it's growing to the point where it doesn't justify the complexity.

Instead, I'm moving the check into Wasmer's instantiation, since we construct a Wasm instance right away anyway, and it takes care of zero-filling and loading all the data segments into memory.

This makes the check more reliable and simpler, as well as avoids double-parsing of Wasm, but it does mean that we'll need to copy it to other hosts in the unlikely event we want to support 2 Wasm hosts simultaneously in the future.

IMO in this case the tradeoff is worth it.

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

### CODE REVIEW

File: crates/core/src/host/wasmer/wasmer_module.rs [hunk 1]
```diff
@@ -174,6 +175,37 @@ impl module_host_actor::WasmInstancePre for WasmerModule {
             .map_err(|err| InitializationError::Instantiation(err.into()))?;
 
         let mem = Mem::extract(&instance.exports).unwrap();
+
+        let abi_version = instance
```
Comment (kulakowski) [inline crates/core/src/host/wasmer/wasmer_module.rs:179]: I would maybe just put some of the (very good!) explanation of the choice to do this at instantiation time in the commit message here as a comment too.

I agree that writing your own increasingly elaborate abstract interpreter to read an int is not worth failing slightly faster.

**Review by kulakowski — APPROVED** [review]

File: crates/core/src/host/wasmer/wasmer_module.rs (hunk 1 above)
Comment (cloutiertyler) [reply crates/core/src/host/wasmer/wasmer_module.rs:179]: +1 on the comment.

**Comment by cloutiertyler** [comment]
@RReverser Can you add your PR comment to the code in some form?

@coolreader18 Can you familiarize yourself with this PR and let me know if it's okay with you?

**Comment by coolreader18** [comment]
Oh, sorry I missed this PR before now.

Hmmm, idk. The reason I did it pre-instantiation is because it provides a clearer and more reliable error about abi mismatches than the more likely error that the module failed to link due to missing imports or something. And yeah, could maybe just do versioned namespacing like wasi does, but this provides a more robust versioning mechanism with the minor component allowing for non-breaking additions.

What about like, a custom section?

**Comment by RReverser** [comment]
Hm that's a good point.

> What about like, a custom section?

I suspect it might be stripped by many toolchains, and some likely have no way of generating one in the first place, so I'm not sure it's future-proof.

> And yeah, could maybe just do versioned namespacing like wasi does, but this provides a more robust versioning mechanism with the minor component allowing for non-breaking additions.

I actually like this idea. I think major-based namespacing like WASI does still allows for non-breaking additions - it's just that you'd get missing import error if you tried to run a new module on old runtime, but that should nearly never happen anyway? And we can provide a nice user-facing error for that scenario as it's easy to detect when import is missing but major version matches.

**Comment by coolreader18** [comment]
Oh actually, yeah, that would work. Maybe this isn't exactly what you were thinking, but it could be like

```wat
(import "spacetime_5.0" "insert" (func ..) $stdb_insert)
(import "spacetime_5.1" "cool_new_function" (func ..) $stdb_cool_new_function)
```

'cause:

> you'd get missing import error if you tried to run a new module on old runtime, but that should nearly never happen anyway

I guess this was something I was hoping to have a nice error about - being able to say exactly what abi version the host will have to supply is something that's I think valuable, and something that can avoid confusion with maybe self-hosted stuff. Cause the minor version is gonna be what's actually changing once the abi stabilizes more.

**Comment by RReverser** [comment]
> Maybe this isn't exactly what you were thinking, but it could be like

Hmm yeah I was rather thinking about keeping just the major version like WASI does, but then iterating over [Module::imports](https://docs.rs/wasmer/latest/wasmer/struct.Module.html#method.imports) ourselves before instantiating the module. That way we can detect when a module uses an import we don't yet have, which would be the indicator that module is built against a newer minor ABI.

You're right that it doesn't give the exact ABI version though. For that your proposed system would be better, but, as a tradeoff, requires more code updates on version bump as it's no longer a single find-replace string.

**Comment by RReverser** [comment]
It almost makes we wonder if, instead of a centralised ABI version, we should have versions on each function individually.

That way module that uses a subset of functions that haven't changed their signatures / behaviour in months would be compatible with a wide range of versions without any changes whatsoever, and, whenever we break some ABI, only those functions that do get modified need to have their version updated in code.

**Comment by coolreader18** [comment]
That does sound like it'd have some benefits, but then that's either locking us into supporting all functions forever, or having a much more confusing set of versions and version support.

**Comment by RReverser** [comment]
Ok I'm open to any of the proposed solutions above if you want to take a stab at this.

**Comment by RReverser** [comment]
@coolreader18 Are you ok with landing this as-is for now and making future improvements later? I just want to fix the issue with check breaking on wasm-opt binaries for now.

**Review by cloutiertyler — APPROVED** [review]
I'm approving this but we should revisit in #335

**Pull request merged**